proposals/CloudIdentifier.md

```diff
@@ -0,0 +1,197 @@
+  "path": "C:\\Program Files\\FooCompany\\cloud_storage_provider_helper.exe",
+  "synced_paths": [
+    "C:\\path_to_synced_directory\\",
+    "C:\\path_to_other_synced_directory\\",
+  ]
+}
+```
+
+#### Requesting a cloud identifier
+
+Whenever a web app calls `getCloudIdentifiers()`, the browser will iterate through all registered CSP clients and filter for ones that cover the file’s/directory’s path as part of their `synced_paths`.
+
+For these, the browser will launch the executable at `path` and transfer the web app’s origin and the requested file/directory path, either via command line argument or via stdin (preceded with a 32-bit message length).
+
+The CSP client will then respond with the token for the given file/directory via stdout (preceded with a 32-bit message length).
+
+Once the browser has received all responses (or some reasonable timeout has expired), the browser will bundle the individual tokens and the CSPs’ names together to return a cloud identifier.
+
+It is up to each individual browser’s implementation on how they handle dynamic registrations, i.e. whether they re-read all the CSP registration files for each request and how long or if they cache them.
+
+## Security and Privacy Considerations
+
+### Fingerprinting
+
+The browser has no control whether the CSP will provide one-time tokens or stable identifiers as part of their cloud identifier. If the CSP provides stable identifiers, the web application could use these as a [fingerprinting](https://www.w3.org/TR/fingerprinting-guidance/#dfn-active-fingerprinting) mechanism for the files/directories it has access to.
+
+In theory, if a web application already has access to a `FileSystemHandle` the web application could already use other mechanisms to generate fingerprinting identifiers, but these are less stable:
+* Hashing the file’s/directory’s contents -> identifier will change if file/directory content changes
+* `FileSystemHandle::getUniqueId()` [[explainer](https://github.com/whatwg/fs/pull/46)] -> clearing browsing data will reset unique IDs
+
+The web application would need repeated access to the same `FileSystemHandle` to perform this fingerprinting though. That means the user must either re-grant access to the same file or the web app stores the file handle in an IndexDB, which can be cleared by the user though by clearing their browsing data.
+
+It would also be up to the CSP whether they actually provide permanent tokens or temporary tokens.
+
+### Modification via read-only permission
+
+If a web application only has `read` level access to a `FileSystemHandle`, but has `write` level access to that file via CSP web APIs, it could still modify the cloud-stored file, which is then synced to the device and thereby modify the file. 
+
+In this case, the user has clearly granted write access to the CSP-backed file and by having a sync client, the user also allows the local files to be modified, so the change would actually not be surprising.
```

Comment:
I would emphasize that this is possible today with existing APIs. From the perspective of the web platform APIs, the website is just writing to a standard file on disk and it just so happens that the changes are being synced by the CSP

Reply:
replaced `readwrite` with "edit/write".

That is pretty much out of scope of this web API. It could very well be that the web app only has read permission on the CSP backend level. Not all use cases involve actually writing to the CSP backend.
Line 95 describes this delegation of getting the correct access levels.